copilot-theorem/src/Copilot/Theorem/What4/Translate.hs

```diff
@@ -645,39 +716,96 @@ translateOp2 :: forall sym a b c.
    CE.Op2 a b c ->
    XExpr sym ->
    XExpr sym ->
-   IO (XExpr sym)
+   TransM sym (XExpr sym)
 translateOp2 origExpr sym op xe1 xe2 = case (op, xe1, xe2) of
-  (CE.And, XBool e1, XBool e2) -> XBool <$> WI.andPred sym e1 e2
-  (CE.Or, XBool e1, XBool e2) -> XBool <$> WI.orPred sym e1 e2
+  (CE.And, XBool e1, XBool e2) -> liftIO $ XBool <$> WI.andPred sym e1 e2
+  (CE.Or, XBool e1, XBool e2) -> liftIO $ XBool <$> WI.orPred sym e1 e2
   (CE.And, _, _) -> unexpectedValues "and operation"
   (CE.Or, _, _) -> unexpectedValues "or operation"
-  (CE.Add _, xe1, xe2) -> numOp (WI.bvAdd sym)
-                                (\(_ :: WFP.FloatInfoRepr fi) -> WFP.iFloatAdd @_ @fi sym fpRM)
-                                xe1 xe2
-  (CE.Sub _, xe1, xe2) -> numOp (WI.bvSub sym)
-                                (\(_ :: WFP.FloatInfoRepr fi) -> WFP.iFloatSub @_ @fi sym fpRM)
-                                xe1 xe2
-  (CE.Mul _, xe1, xe2) -> numOp (WI.bvMul sym)
-                                (\(_ :: WFP.FloatInfoRepr fi) -> WFP.iFloatMul @_ @fi sym fpRM)
-                                xe1 xe2
-  (CE.Mod _, xe1, xe2) -> bvOp (WI.bvSrem sym) (WI.bvUrem sym) xe1 xe2
-  (CE.Div _, xe1, xe2) -> bvOp (WI.bvSdiv sym) (WI.bvUdiv sym) xe1 xe2
-  (CE.Fdiv _, xe1, xe2) -> fpOp (\(_ :: WFP.FloatInfoRepr fi) -> WFP.iFloatDiv @_ @fi sym fpRM)
-                                xe1 xe2
-  (CE.Pow _, xe1, xe2) -> fpSpecialOp WSF.Pow xe1 xe2
-  (CE.Logb _, xe1, xe2) -> fpOp logbFn xe1 xe2
+  (CE.Add _, xe1, xe2) -> do addBVSidePred2 xe1 xe2 $ \e1 e2 _ sgn ->
+                               -- The arguments should not result in
+                               -- overflow or underflow
+                               case sgn of
+                                 Signed -> do
+                                   (wrap, _) <- WI.addSignedOF sym e1 e2
+                                   WI.notPred sym wrap
+                                 Unsigned -> do
+                                   (wrap, _) <- WI.addUnsignedOF sym e1 e2
+                                   WI.notPred sym wrap
```

Comment:
This adds side conditions for both signed and unsigned integer overflow. That being said, I haven't quite figured out how to get Clang to emit `add` instructions with the `nuw` flag (indicating that unsigned wrapping should result in poison) when the optimization level is `-O0`. I decided to include the `Unsigned` case anyway for symmetry with the `Signed` case, but I'm undecided on if we should keep that design.

Reply:
Oh, wait, actually... I misunderstood.  This is for the copilot semantics.  I guess it's an open question what the semantics of unsigned overflow should be in Copilot.  My usual stance is that unsigned arithmetic is actually just arithmetic modulo `2^n`, so overflow is fine and should not be an error condition.

Reply:
I've change this code so that only signed overflow/underflow results in an error condition being emitted.